### PR TITLE
Version 4

### PR DESCRIPTION
**Pull request by Jasonej**
Version 4 seeks to make use of new functionality provided by PHP 8.0 and more strongly focus on formatting and a simpler, more extensible enhancement mechanism. 

Changes without clutter from the clean slate commit can be found [here](https://github.com/sourcetoad/enhanced-resources/pull/25/files/7ac1a3c2e478023f82cecfd14b395a4a582e8f2c..f830c19c45402b6b26245d180854c1f594483205).

### Formatting
- [x] Drop requirement for a format method naming convention.
- [x] Support format aliases.
- [x] Explicitly defined defaults.
- [x] Implicit default when only format is present.
- [x] Allow for selection of specific formats.
- [x] Explicit default format on child class overrides default from parent class.
- [x] Integrate formatting with Resource.
- [x] Integrate formatting with ResourceCollection.
- [x] Integrate formatting with AnonymousResourceCollection.
### Enhancements
- [x] Stack based enhancement application (similar to how state is applied with Eloquent factories).
- [x] Powered by a `modify` method.
- [x] Trait for the `except` enhancement.
- [x] Trait for the `only` enhancement.
- [x] Integrate enhancements with ResourceCollection.
- [x] Integrate enhancements with AnonymousResourceCollection.
### Miscellaneous
- [x] Support specifying the intended response status code.
- [x] Resourceable trait.
- [x] Documentation.

## Notes
- PHP 8.0 is required.
- The `append` enhancement is being dropped.
- The `call` enhancement is being replaced by the core `enhance` method functionality which will power the enhancement mechanism.
- The `replace` enhancement is being dropped.

### CODE REVIEW

**Comment by erik-perri** [comment]
Do any of the enhancement changes open the option to hook into the resource generation globally?

As an example, I'd like to force every resource in a project to return an automatically generated `meta.type` property.

**Comment by Jasonej** [comment]
> Do any of the enhancement changes open the option to hook into the resource generation globally?
> 
> As an example, I'd like to force every resource in a project to return an automatically generated `meta.type` property.

I hadn't considered doing anything to support global, but I have been considering allowing for formats/enhancements for meta data. The with/additional data system provided by Laravel's resources is a little less straightforward, and the relationship between collections and meta data is something that would require some thought if we want to support enhancing it.

Do you have an idea for how you'd want to apply these global enhancements?

**Comment by erik-perri** [comment]
My initial thought was having it function like View composers, but the more I have thought about it the less sure I am about it being a useful feature.

I have been trying to think of other data that makes sense to attach globally and have not though of anything else that wouldn’t require a big `if instanceof` block, which is probably not something we want to encourage.

I will think about it some more to see if I can come up with anything else that could sanely make use of global hooks or a better approach to applying something like generic meta type data to each resource.

**Comment by erik-perri** [comment]
I am unable to install the alpha due to the old service provider reference in `composer.json`

```console
$ composer require sourcetoad/enhanced-resources:v4.0.0-alpha1
./composer.json has been updated
Running composer update sourcetoad/enhanced-resources
Loading composer repositories with package information
Updating dependencies
Lock file operations: 1 install, 0 updates, 0 removals
  - Locking sourcetoad/enhanced-resources (v4.0.0-alpha1)
Writing lock file
Installing dependencies from lock file (including require-dev)
Package operations: 1 install, 0 updates, 0 removals
  - Installing sourcetoad/enhanced-resources (v4.0.0-alpha1): Extracting archive
Generating optimized autoload files
> Illuminate\Foundation\ComposerScripts::postAutoloadDump
> @php artisan package:discover --ansi

In ProviderRepository.php line 208:

  Class "Sourcetoad\EnhancedResources\ServiceProvider" not found


Script @php artisan package:discover --ansi handling the post-autoload-dump event returned with error code 1

Installation failed, reverting ./composer.json and ./composer.lock to their original content.
```

**Comment by Jasonej** [comment]
> I am unable to install the alpha due to the old service provider reference in `composer.json`

Fixed and retagged.

**Review by joeladam518 — DISMISSED** [review]

**Review by erik-perri — DISMISSED** [review]

**Review by erik-perri — DISMISSED** [review]

**Review by joeladam518 — DISMISSED** [review]

**Review by NilsDelaguardia — DISMISSED** [review]

**Review by frankspress — DISMISSED** [review]

File: README.md [hunk 1]
```diff
@@ -10,146 +10,312 @@ $ composer require sourcetoad/enhanced-resources
...
-### Collections
+```php
+<?php
 
-Enhanced collections work by mapping the format and enhancement calls to each resource contained within.
-As a result enhanced collections only work when they collect enhanced resources.
-The easiest way to handle this is to set the `$collects` property on the enhanced collection or ensure that you're
-following the standard convention for resource and collection naming so that the `collects()` method can detect it.
+use Sourcetoad\EnhancedResources\Enhancements\Only;
+use Sourcetoad\EnhancedResources\Enhancements\Traits\HasExceptEnhancement;
```
Comment (jyann) [inline README.md]: Should this be `HasOnlyEnhancement`?

Comment (Jasonej) [reply README.md]: Fixed, thanks!

Comment (jyann) [reply README.md]: Sure. Excited about all this new stuff! I like how the features are becoming more explicit and transparent

**Review by jyann — DISMISSED** [review]

**Comment by erik-perri** [comment]
What is your opinion on resolving the `AnonymousResourceCollection` instance when creating a collection rather than using `new`?

In one of my projects, I previously created a custom collection resource handler to return the pagination meta information in a format I prefer.  I decided to switch to the `::collection` pattern using the pagination overrides Laravel now provides (see [this](https://github.com/laravel/framework/pull/39600/commits/bb9909bbb4ed291bf495a6f125228cb7153bb7b0) PR).   Doing so requires me to create both a `Resource` override with a `collection` method, and an `AnonymousResourceCollection` with the `paginationInformation` method.  If it were resolved instead I could just create the `AnonymousResourceCollection` and use the container to switch the implementation.  Not an issue if you'd prefer to leave it, one additional base class is not a big concern.

**Review by NilsDelaguardia — DISMISSED** [review]

**Comment by Jasonej** [comment]
> What is your opinion on resolving the `AnonymousResourceCollection` instance when creating a collection rather than using `new`?

Resolving it makes a ton of sense. I'll look into adding that in a bit.

**Comment by Jasonej** [comment]
Added Laravel 9 support, resolving AnonymousResourceCollection instead of direct instantiation, and tagged as v4.0.0-rc1.

**Review by erik-perri — APPROVED** [review]

**Comment by joeladam518** [comment]
This might need to be pushed to v5 because of #26. Adding Laravel 9 support required me to drop PHP7.4 and Laravel 7 support.

**Pull request closed without merging**